pkg/sentry/socket/netlink/route/protocol.go

```diff
@@ -500,6 +500,79 @@ func (p *Protocol) delAddr(ctx context.Context, msg *netlink.Message, ms *netlin
 	return nil
 }
 
+func fetchRoute(msg *netlink.Message) (*inet.Route, *syserr.Error) {
+	var rtMsg linux.RouteMessage
+	attrs, ok := msg.GetData(&rtMsg)
+	if !ok {
+		return nil, syserr.ErrInvalidArgument
+	}
+	if rtMsg.Table != linux.RT_TABLE_MAIN {
+		// Only the main table since we don't have multiple
+		// routing tables.
+		return nil, syserr.ErrInvalidArgument
+	}
+	route := inet.Route{
+		Family:   rtMsg.Family,
+		DstLen:   rtMsg.DstLen,
+		SrcLen:   rtMsg.SrcLen,
+		TOS:      rtMsg.TOS,
+		Table:    rtMsg.Table,
+		Protocol: rtMsg.Protocol,
+		Scope:    rtMsg.Scope,
+		Type:     rtMsg.Type,
+		Flags:    rtMsg.Flags,
+	}
+
+	attHdr, buff, rest, ok := attrs.ParseFirst()
+	for ok {
+		switch attHdr.Type {
+		case linux.RTA_DST:
+			route.DstAddr = buff
+		case linux.RTA_SRC:
+			route.SrcAddr = buff
+		case linux.RTA_GATEWAY:
+			route.GatewayAddr = buff
+		case linux.RTA_OIF:
+			// This conversion is also present in the kernel...
+			var outputIF primitive.Int32
+			outputIF.UnmarshalUnsafe(buff)
+			route.OutputInterface = int32(outputIF)
+		default:
+			// Other attributes are not implemented yet.
+			return nil, syserr.ErrInvalidArgument
+		}
+		attHdr, buff, rest, ok = rest.ParseFirst()
+	}
+	return &route, nil
+}
+
+// newRoute handles RTM_NEWROUTE requests.
+func (p *Protocol) newRoute(ctx context.Context, msg *netlink.Message, ms *netlink.MessageSet) *syserr.Error {
+	stack := inet.StackFromContext(ctx)
+	if stack == nil {
+		// No network routes.
+		return nil
+	}
+	hdr := msg.Header()
+	flags := hdr.Flags&^(linux.NLM_F_REQUEST|linux.NLM_F_ACK)
+	if flags != linux.NLM_F_CREATE|linux.NLM_F_EXCL {
+		// ip route prepend/append/change/replace/test are not implemented.
+		return syserr.ErrInvalidArgument
+	}
+	route, sysErr := fetchRoute(msg)
+	if sysErr != nil {
+		return sysErr
+	}
+	err := stack.AddRoute(*route)
+	if err != nil {
+		if err.Error() == syserr.ErrExists.String() {
+			return syserr.ErrExists
+		}
+		return syserr.ErrInvalidArgument
+	}
```

Comment:
I think you can just return the error now since it's been translated already.

Reply:
Unfortunately no, because type the function returns type `*syserr.Error` which does not implement `error`.

Reply:
Ah, yeah. In any case you'll want to do this as a type assertion rather than comparing the error messages.

Reply:
Fixed :)